File: maubot/management/frontend/src/pages/dashboard/InstanceDatabase.js
// maubot - A plugin-based Matrix bot system.
// Copyright (C) 2021 Tulir Asokan
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
import React, { Component } from "react"
import { NavLink, Link, withRouter } from "react-router-dom"
import { ContextMenu, ContextMenuTrigger, MenuItem } from "react-contextmenu"
import { ReactComponent as ChevronLeft } from "../../res/chevron-left.svg"
import { ReactComponent as OrderDesc } from "../../res/sort-down.svg"
import { ReactComponent as OrderAsc } from "../../res/sort-up.svg"
import api from "../../api"
import Spinner from "../../components/Spinner"

// eslint-disable-next-line no-extend-native
Map.prototype.map = function(func) {
    const res = []
    for (const [key, value] of this) {
        res.push(func(value, key, this))
    }
    return res
}

class InstanceDatabase extends Component {
    constructor(props) {
        super(props)
        this.state = {
            tables: null,
            header: null,
            content: null,
            query: "",
            selectedTable: null,

            error: null,

            prevQuery: null,
            rowCount: null,
            insertedPrimaryKey: null,
        }
        this.order = new Map()
    }

    async componentDidMount() {
        const tables = new Map(Object.entries(await api.getInstanceDatabase(this.props.instanceID)))
        for (const [name, table] of tables) {
            table.name = name
            table.columns = new Map(Object.entries(table.columns))
            for (const [columnName, column] of table.columns) {
                column.name = columnName
                column.sort = null
            }
        }
        this.setState({ tables })
        this.checkLocationTable()
    }

    componentDidUpdate(prevProps) {
        if (this.props.location !== prevProps.location) {
            this.order = new Map()
            this.setState({ header: null, content: null })
            this.checkLocationTable()
        }
    }

    checkLocationTable() {
        const prefix = `/instance/${this.props.instanceID}/database/`
        if (this.props.location.pathname.startsWith(prefix)) {
            const table = this.props.location.pathname.substr(prefix.length)
            this.setState({ selectedTable: table })
            this.buildSQLQuery(table)
        }
    }

    getSortQueryParams(table) {
        const order = []
        for (const [column, sort] of Array.from(this.order.entries()).reverse()) {
            order.push(`order=${column}:${sort}`)
        }
        return order
    }

    buildSQLQuery(table = this.state.selectedTable, resetContent = true) {
        let query = `SELECT * FROM "${table}"`

        if (this.order.size > 0) {
            const order = Array.from(this.order.entries()).reverse()
                .map(([column, sort]) => `${column} ${sort}`)
            query += ` ORDER BY ${order.join(", ")}`
        }

        query += " LIMIT 100"
        this.setState({ query }, () => this.reloadContent(resetContent))
    }

    reloadContent = async (resetContent = true) => {
        this.setState({ loading: true })
        const res = await api.queryInstanceDatabase(this.props.instanceID, this.state.query)
        this.setState({ loading: false })
        if (resetContent) {
            this.setState({
                prevQuery: null,
                rowCount: null,
                insertedPrimaryKey: null,
                error: null,
            })
        }
        if (!res.ok) {
            this.setState({
                error: res.error,
            })
        } else if (res.rows) {
            this.setState({
                header: res.columns,
                content: res.rows,
            })
        } else {
            this.setState({
                prevQuery: res.query,
                rowCount: res.rowcount,
                insertedPrimaryKey: res.insertedPrimaryKey,
            })
            this.buildSQLQuery(this.state.selectedTable, false)
        }
    }

    toggleSort(column) {
        const oldSort = this.order.get(column) || "auto"
        this.order.delete(column)
        switch (oldSort) {
        case "auto":
            this.order.set(column, "DESC")
            break
        case "DESC":
            this.order.set(column, "ASC")
            break
        case "ASC":
        default:
            break
        }
        this.buildSQLQuery()
    }

    getSortIcon(column) {
        switch (this.order.get(column)) {
        case "DESC":
            return <OrderDesc/>
        case "ASC":
            return <OrderAsc/>
        default:
            return null
        }
    }

    getColumnInfo(columnName) {
        const table = this.state.tables.get(this.state.selectedTable)
        if (!table) {
            return null
        }
        const column = table.columns.get(columnName)
        if (!column) {
            return null
        }
        if (column.primary) {
            return <span className="meta">&nbsp;(pk)</span>
        } else if (column.unique) {
            return <span className="meta">&nbsp;(u)</span>
        }
        return null
    }

    getColumnType(columnName) {
        const table = this.state.tables.get(this.state.selectedTable)
        if (!table) {
            return null
        }
        const column = table.columns.get(columnName)
        if (!column) {
            return null
        }
        return column.type
    }

    deleteRow = async (_, data) => {
        const values = this.state.content[data.row]
        const keys = this.state.header
        const condition = []
        for (const [index, key] of Object.entries(keys)) {
            const val = values[index]
            condition.push(`${key}='${this.sqlEscape(val.toString())}'`)
        }
        const query = `DELETE FROM "${this.state.selectedTable}" WHERE ${condition.join(" AND ")}`
        const res = await api.queryInstanceDatabase(this.props.instanceID, query)
        this.setState({
            prevQuery: `DELETE FROM "${this.state.selectedTable}" ...`,
            rowCount: res.rowcount,
        })
        await this.reloadContent(false)
    }

    editCell = async (evt, data) => {
        console.log("Edit", data)
    }

    collectContextMeta = props => ({
        row: props.row,
        col: props.col,
    })

    // eslint-disable-next-line no-control-regex
    sqlEscape = str => str.replace(/[\0\x08\x09\x1a\n\r"'\\%]/g, char => {
        switch (char) {
        case "\0":
            return "\\0"
        case "\x08":
            return "\\b"
        case "\x09":
            return "\\t"
        case "\x1a":
            return "\\z"
        case "\n":
            return "\\n"
        case "\r":
            return "\\r"
        case "\"":
        case "'":
        case "\\":
        case "%":
            return "\\" + char
        default:
            return char
        }
    })

    renderTable = () => <div className="table">
        {this.state.header ? <>
            <table>
                <thead>
                    <tr>
                        {this.state.header.map(column => (
                            <td key={column}>
                                <span onClick={() => this.toggleSort(column)}
                                      title={this.getColumnType(column)}>
                                    <strong>{column}</strong>
                                    {this.getColumnInfo(column)}
                                    {this.getSortIcon(column)}
                                </span>
                            </td>
                        ))}
                    </tr>
                </thead>
                <tbody>
                    {this.state.content.map((row, rowIndex) => (
                        <tr key={rowIndex}>
                            {row.map((cell, colIndex) => (
                                <ContextMenuTrigger key={colIndex} id="database_table_menu"
                                                    renderTag="td" row={rowIndex} col={colIndex}
                                                    collect={this.collectContextMeta}>
                                    {cell}
                                </ContextMenuTrigger>
                            ))}
                        </tr>
                    ))}
                </tbody>
            </table>
            <ContextMenu id="database_table_menu">
                <MenuItem onClick={this.deleteRow}>Delete row</MenuItem>
                <MenuItem disabled onClick={this.editCell}>Edit cell</MenuItem>
            </ContextMenu>
        </> : this.state.loading ? <Spinner/> : null}
    </div>

    renderContent() {
        return <>
            <div className="tables">
                {this.state.tables.map((_, tbl) => (
                    <NavLink key={tbl} to={`/instance/${this.props.instanceID}/database/${tbl}`}>
                        {tbl}
                    </NavLink>
                ))}
            </div>
            <div className="query">
                <input type="text" value={this.state.query} name="query"
                       onChange={evt => this.setState({ query: evt.target.value })}/>
                <button type="submit" onClick={this.reloadContent}>Query</button>
            </div>
            {this.state.error && <div className="error">
                {this.state.error}
            </div>}
            {this.state.prevQuery && <div className="prev-query">
                <p>
                    Executed <span className="query">{this.state.prevQuery}</span> -
                    affected <strong>{this.state.rowCount} rows</strong>.
                </p>
                {this.state.insertedPrimaryKey && <p className="inserted-primary-key">
                    Inserted primary key: {this.state.insertedPrimaryKey}
                </p>}
            </div>}
            {this.renderTable()}
        </>
    }

    render() {
        return <div className="instance-database">
            <div className="topbar">
                <Link className="topbar" to={`/instance/${this.props.instanceID}`}>
                    <ChevronLeft/>
                    Back
                </Link>
            </div>
            {this.state.tables
                ? this.renderContent()
                : <Spinner className="maubot-loading"/>}
        </div>
    }
}

export default withRouter(InstanceDatabase)

File: maubot/management/api/instance_database.py
# maubot - A plugin-based Matrix bot system.
# Copyright (C) 2022 Tulir Asokan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
from __future__ import annotations

from datetime import datetime

from aiohttp import web
from asyncpg import PostgresError
from sqlalchemy import asc, desc, engine, exc
import aiosqlite

from mautrix.util.async_db import Database

from ...instance import PluginInstance
from .base import routes
from .responses import resp


@routes.get("/instance/{id}/database")
async def get_database(request: web.Request) -> web.Response:
    instance_id = request.match_info["id"].lower()
    instance = await PluginInstance.get(instance_id)
    if not instance:
        return resp.instance_not_found
    elif not instance.inst_db:
        return resp.plugin_has_no_database
    return web.json_response(await instance.get_db_tables())


@routes.get("/instance/{id}/database/{table}")
async def get_table(request: web.Request) -> web.Response:
    instance_id = request.match_info["id"].lower()
    instance = await PluginInstance.get(instance_id)
    if not instance:
        return resp.instance_not_found
    elif not instance.inst_db:
        return resp.plugin_has_no_database
    tables = await instance.get_db_tables()
    try:
        table = tables[request.match_info.get("table", "")]
    except KeyError:
        return resp.table_not_found
    try:
        order = [tuple(order.split(":")) for order in request.query.getall("order")]
        order = [
            (asc if sort.lower() == "asc" else desc)(table.columns[column])
            if sort
            else table.columns[column]
            for column, sort in order
        ]
    except KeyError:
        order = []
    limit = int(request.query.get("limit", "100"))
    if isinstance(instance.inst_db, engine.Engine):
        return _execute_query_sqlalchemy(instance, table.select().order_by(*order).limit(limit))


@routes.post("/instance/{id}/database/query")
async def query(request: web.Request) -> web.Response:
    instance_id = request.match_info["id"].lower()
    instance = await PluginInstance.get(instance_id)
    if not instance:
        return resp.instance_not_found
    elif not instance.inst_db:
        return resp.plugin_has_no_database
    data = await request.json()
    try:
        sql_query = data["query"]
    except KeyError:
        return resp.query_missing
    rows_as_dict = data.get("rows_as_dict", False)
    if isinstance(instance.inst_db, engine.Engine):
        return _execute_query_sqlalchemy(instance, sql_query, rows_as_dict)
    elif isinstance(instance.inst_db, Database):
        try:
            return await _execute_query_asyncpg(instance, sql_query, rows_as_dict)
        except (PostgresError, aiosqlite.Error) as e:
            return resp.sql_error(e, sql_query)
    else:
        return resp.unsupported_plugin_database


def check_type(val):
    if isinstance(val, datetime):
        return val.isoformat()
    return val


async def _execute_query_asyncpg(
    instance: PluginInstance, sql_query: str, rows_as_dict: bool = False
) -> web.Response:
    data = {"ok": True, "query": sql_query}
    if sql_query.upper().startswith("SELECT"):
        res = await instance.inst_db.fetch(sql_query)
        data["rows"] = [
            (
                {key: check_type(value) for key, value in row.items()}
                if rows_as_dict
                else [check_type(value) for value in row]
            )
            for row in res
        ]
        if len(res) > 0:
            # TODO can we find column names when there are no rows?
            data["columns"] = list(res[0].keys())
    else:
        res = await instance.inst_db.execute(sql_query)
        if isinstance(res, str):
            data["status_msg"] = res
        elif isinstance(res, aiosqlite.Cursor):
            data["rowcount"] = res.rowcount
            # data["inserted_primary_key"] = res.lastrowid
        else:
            data["status_msg"] = "unknown status"
    return web.json_response(data)


def _execute_query_sqlalchemy(
    instance: PluginInstance, sql_query: str, rows_as_dict: bool = False
) -> web.Response:
    assert isinstance(instance.inst_db, engine.Engine)
    try:
        res = instance.inst_db.execute(sql_query)
    except exc.IntegrityError as e:
        return resp.sql_integrity_error(e, sql_query)
    except exc.OperationalError as e:
        return resp.sql_operational_error(e, sql_query)
    data = {
        "ok": True,
        "query": str(sql_query),
    }
    if res.returns_rows:
        data["rows"] = [
            (
                {key: check_type(value) for key, value in row.items()}
                if rows_as_dict
                else [check_type(value) for value in row]
            )
            for row in res
        ]
        data["columns"] = res.keys()
    else:
        data["rowcount"] = res.rowcount
    if res.is_insert:
        data["inserted_primary_key"] = res.inserted_primary_key
    return web.json_response(data)
